test/micromamba/test_shell.py

```diff
@@ -145,20 +156,9 @@ def custom_shell(shell):
     @pytest.mark.parametrize("expanded_home", [False, True])
     @pytest.mark.parametrize("prefix_type", ["prefix", "name"])
     def test_activate(self, shell_type, root, env_exists, prefix_type, expanded_home):
-        if (
-            (platform.system() == "Linux" and shell_type not in ("bash", "posix"))
-            or (
-                platform.system() == "Windows"
-                and shell_type not in ("cmd.exe", "powershell")
-            )
-            or (
-                platform.system() == "Darwin"
-                and shell_type not in ("zsh", "bash", "posix")
-            )
-        ):
-            pytest.skip("Incompatible shell/OS")
+        skip_if_shell_incompat(shell_type)
 
-        if not root and env_exists:
+        if not root and not env_exists:
```

Comment:
Unrelated to the present issue, but obviously went undetected!

Reply:
Hi @ashwinvis 
I was intended that way to test both cases where the env exists ant *not* exists.

You should either keep it and merge with `test_activate_non_existent` you added:
```
if env_exists:
    res = shell(*cmd)
    ...
else:
    with pytest.raises(subprocess.CalledProcessError):
        shell(*cmd)
    return
```
OR
remove the parametrization `@pytest.mark.parametrize("env_exists", [False, True])`

Note that we cover more cases with the parametrized test, so I prefer first option

Reply:
I see! Good point.

Reply:
But I don't understand the logic behind this case

```py
root = False
env_exists = True
if not root and env_exists:
      create("-n", TestShell.env_name, "-q", "--offline", no_dry_run=True)
```

Why would one try to create an environment which already exists?

Reply:
Aah ... I think get it now. You are creating the environment for that test, *so that it exists*

Reply:
yes :)

Reply:
@adriendelsalle I made this change to merge the tests:

```diff
diff --git a/test/micromamba/test_shell.py b/test/micromamba/test_shell.py
index fa56542..88ecd54 100644
--- a/test/micromamba/test_shell.py
+++ b/test/micromamba/test_shell.py
@@ -158,7 +158,8 @@ class TestShell:
     def test_activate(self, shell_type, root, env_exists, prefix_type, expanded_home):
         skip_if_shell_incompat(shell_type)
 
-        if not root and not env_exists:
+        if not root and env_exists:
+            # Create the environment for this test, so that it exists
             create("-n", TestShell.env_name, "-q", "--offline", no_dry_run=True)
 
         if prefix_type == "prefix":
@@ -175,7 +176,12 @@ class TestShell:
         else:
             cmd = ("activate", "-s", shell_type, "-p", TestShell.env_name)
 
-        res = shell(*cmd)
+        if env_exists:
+            res = shell(*cmd)
+        else:
+            with pytest.raises(subprocess.CalledProcessError):
+                shell(*cmd)
+            return
 
         # TODO: improve this test
         assert res
@@ -185,14 +191,6 @@ class TestShell:
             assert f"export CONDA_DEFAULT_ENV='{TestShell.env_name}'" in res
             assert f"export CONDA_PROMPT_MODIFIER='({TestShell.env_name}) '" in res
 
-    @pytest.mark.parametrize("shell_type", ["bash", "posix", "powershell", "cmd.exe"])
-    def test_activate_non_existent(self, shell_type):
-        skip_if_shell_incompat(shell_type)
-
-        cmd = ("activate", "-s", shell_type, "-p", "this-env-does-not-exist")
-        with pytest.raises(subprocess.CalledProcessError):
-            shell(*cmd)
-
     @pytest.mark.parametrize("shell_type", ["bash", "powershell", "cmd.exe"])
     @pytest.mark.parametrize("prefix_selector", [None, "prefix"])
     def test_init(self, shell_type, prefix_selector):

```

and now I get 12 test failures like:

```py
    @pytest.mark.parametrize("shell_type", ["bash", "posix", "powershell", "cmd.exe"])
    @pytest.mark.parametrize("root", [False, True])
    @pytest.mark.parametrize("env_exists", [False, True])
    @pytest.mark.parametrize("expanded_home", [False, True])
    @pytest.mark.parametrize("prefix_type", ["prefix", "name"])
    def test_activate(self, shell_type, root, env_exists, prefix_type, expanded_home):
        skip_if_shell_incompat(shell_type)
    
        if not root and env_exists:
            # Create the environment for this test, so that it exists
            create("-n", TestShell.env_name, "-q", "--offline", no_dry_run=True)
    
        if prefix_type == "prefix":
            if expanded_home:
                cmd = ("activate", "-s", shell_type, "-p", TestShell.prefix)
            else:
                cmd = (
                    "activate",
                    "-s",
                    shell_type,
                    "-p",
                    TestShell.prefix.replace(os.path.expanduser("~"), "~"),
                )
        else:
            cmd = ("activate", "-s", shell_type, "-p", TestShell.env_name)
    
        if env_exists:
            res = shell(*cmd)
        else:
            with pytest.raises(subprocess.CalledProcessError):
>               shell(*cmd)
E               Failed: DID NOT RAISE <class 'subprocess.CalledProcessError'>

```

which in summary:

```sh
❯ pytest test/micromamba/test_shell.py | rg '^shell.*' -A1 
shell_type = 'bash', root = False, env_exists = False, prefix_type = 'prefix'
expanded_home = True
--
shell_type = 'posix', root = False, env_exists = False, prefix_type = 'prefix'
expanded_home = True
--
shell_type = 'bash', root = True, env_exists = False, prefix_type = 'prefix'
expanded_home = True
--
shell_type = 'posix', root = True, env_exists = False, prefix_type = 'prefix'
expanded_home = True
--
shell_type = 'bash', root = False, env_exists = False, prefix_type = 'name'
expanded_home = False
--
shell_type = 'posix', root = False, env_exists = False, prefix_type = 'name'
expanded_home = False
--
shell_type = 'bash', root = True, env_exists = False, prefix_type = 'name'
expanded_home = False
--
shell_type = 'posix', root = True, env_exists = False, prefix_type = 'name'
expanded_home = False
--
shell_type = 'bash', root = False, env_exists = False, prefix_type = 'name'
expanded_home = True
--
shell_type = 'posix', root = False, env_exists = False, prefix_type = 'name'
expanded_home = True
--
shell_type = 'bash', root = True, env_exists = False, prefix_type = 'name'
expanded_home = True
--
shell_type = 'posix', root = True, env_exists = False, prefix_type = 'name'
expanded_home = True
```